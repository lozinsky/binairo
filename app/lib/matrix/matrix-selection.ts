import { type Matrix, type MatrixCell } from './matrix';
import { type MatrixLine } from './matrix-line';

export interface MatrixSelectionPosition {
  readonly x: number;
  readonly y: number;
}

export class MatrixSelection {
  readonly #positions: readonly MatrixSelectionPosition[];

  constructor(positions: readonly MatrixSelectionPosition[]) {
    this.#positions = positions;
  }

  static concat(...selections: readonly MatrixSelection[]) {
    const positions: MatrixSelectionPosition[] = [];

    for (const selection of selections) {
      positions.push(...selection.#positions);
    }

    return new this(positions);
  }

  static from<M extends Matrix<MatrixLine<unknown>>>(matrix: M, cells: ReadonlyArray<MatrixCell<M>>) {
    const positions: MatrixSelectionPosition[] = [];

    for (const [y, line] of matrix.entries()) {
      for (const [x, cell] of line.entries()) {
        for (const other of cells) {
          if (cell === other) {
            positions.push({ x, y });
          }
        }
      }
    }

    return new this(positions);
  }

  static parse(value: string) {
    return new this(JSON.parse(atob(value)) as readonly MatrixSelectionPosition[]);
  }

  exclude(other: MatrixSelection) {
    const positions: MatrixSelectionPosition[] = [...this.#positions];

    for (const position of other.#positions) {
      const index = positions.findIndex((other) => other.x === position.x && other.y === position.y);

      if (index === -1) {
        continue;
      }

      positions.splice(index, 1);
    }

    return new MatrixSelection(positions);
  }

  execute<M extends Matrix<MatrixLine<unknown>>>(matrix: M) {
    const cells: Array<MatrixCell<M>> = [];

    for (const position of this.#positions) {
      const line = matrix.at(position.y);

      if (line === undefined) {
        continue;
      }

      const cell = line.at(position.x);

      if (cell === undefined) {
        continue;
      }

      cells.push(cell as MatrixCell<M>);
    }

    return cells;
  }

  toString() {
    return btoa(JSON.stringify(this.valueOf()));
  }

  valueOf() {
    return this.#positions;
  }

  get length() {
    return this.#positions.length;
  }
}
