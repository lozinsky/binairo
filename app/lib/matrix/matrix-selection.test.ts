import { expect, test } from 'vitest';

import { MatrixSelection } from './matrix-selection';

test.each([
  [
    [
      [1, 2, 3],
      [4, 5, 6],
      [7, 8, 9],
    ],
    [1, 1, 4, 5, 7, 9],
  ],
  [
    [
      [1, 2],
      [4, 5],
    ],
    [1, 1, 4, 5],
  ],
])('collects matrix selection from cells', (matrix, cells) => {
  expect(MatrixSelection.collect(matrix, cells).execute(matrix)).toMatchSnapshot();
});

test.each([
  [
    [
      new MatrixSelection([{ x: 0, y: 0 }]),
      new MatrixSelection([{ x: 0, y: 0 }]),
      new MatrixSelection([{ x: 0, y: 1 }]),
      new MatrixSelection([{ x: 1, y: 1 }]),
      new MatrixSelection([{ x: 0, y: 2 }]),
      new MatrixSelection([{ x: 2, y: 2 }]),
    ],
  ],
  [
    [
      new MatrixSelection([{ x: 0, y: 0 }]),
      new MatrixSelection([{ x: 0, y: 0 }]),
      new MatrixSelection([{ x: 0, y: 1 }]),
      new MatrixSelection([{ x: 1, y: 1 }]),
    ],
  ],
])('concatenates matrix selection', (selections) => {
  expect(MatrixSelection.concat(...selections)).toMatchSnapshot();
});

test.each([
  'W3sieCI6MCwieSI6MH0seyJ4IjowLCJ5IjowfSx7IngiOjAsInkiOjF9LHsieCI6MSwieSI6MX0seyJ4IjowLCJ5IjoyfSx7IngiOjIsInkiOjJ9XQ==',
  'W3sieCI6MCwieSI6MH0seyJ4IjowLCJ5IjowfSx7IngiOjAsInkiOjF9LHsieCI6MSwieSI6MX1d',
])('returns matrix selection from string', (value) => {
  expect(MatrixSelection.parse(value)).toMatchSnapshot();
});

test.each([
  [
    new MatrixSelection([
      { x: 0, y: 0 },
      { x: 0, y: 0 },
      { x: 1, y: 1 },
    ]),
    new MatrixSelection([{ x: 0, y: 0 }]),
  ],
  [
    new MatrixSelection([
      { x: 0, y: 0 },
      { x: 0, y: 0 },
      { x: 1, y: 1 },
    ]),
    new MatrixSelection([
      { x: 0, y: 0 },
      { x: 0, y: 0 },
    ]),
  ],
  [
    new MatrixSelection([
      { x: 0, y: 0 },
      { x: 0, y: 0 },
      { x: 1, y: 1 },
    ]),
    new MatrixSelection([
      { x: 0, y: 0 },
      { x: 0, y: 0 },
      { x: 1, y: 1 },
    ]),
  ],
])('excludes matrix selection', (selection, other) => {
  expect(selection.exclude(other)).toMatchSnapshot();
});

test.each([
  [
    [
      [1, 2, 3],
      [4, 5, 6],
      [7, 8, 9],
    ],
    new MatrixSelection([
      { x: 0, y: 0 },
      { x: 0, y: 0 },
      { x: 0, y: 1 },
      { x: 1, y: 1 },
      { x: 0, y: 2 },
      { x: 2, y: 2 },
    ]),
  ],
  [
    [
      [1, 2],
      [4, 5],
    ],
    new MatrixSelection([
      { x: 0, y: 0 },
      { x: 0, y: 0 },
      { x: 0, y: 1 },
      { x: 1, y: 1 },
      { x: 0, y: 2 },
      { x: 2, y: 2 },
    ]),
  ],
])('returns matrix selection cells', (matrix, selection) => {
  expect(selection.execute(matrix)).toMatchSnapshot();
});

test.each([
  new MatrixSelection([
    { x: 0, y: 0 },
    { x: 0, y: 0 },
    { x: 0, y: 1 },
    { x: 1, y: 1 },
    { x: 0, y: 2 },
    { x: 2, y: 2 },
  ]),
  new MatrixSelection([
    { x: 0, y: 0 },
    { x: 0, y: 0 },
    { x: 0, y: 1 },
    { x: 1, y: 1 },
  ]),
])('returns matrix selection string', (selection) => {
  expect(selection.toString()).toMatchSnapshot();
});

test.each([
  new MatrixSelection([
    { x: 0, y: 0 },
    { x: 0, y: 0 },
    { x: 0, y: 1 },
    { x: 1, y: 1 },
    { x: 0, y: 2 },
    { x: 2, y: 2 },
  ]),
  new MatrixSelection([
    { x: 0, y: 0 },
    { x: 0, y: 0 },
    { x: 0, y: 1 },
    { x: 1, y: 1 },
  ]),
])('returns matrix selection value of', (selection) => {
  expect(selection.valueOf()).toMatchSnapshot();
});

test.each([
  new MatrixSelection([
    { x: 0, y: 0 },
    { x: 0, y: 0 },
    { x: 0, y: 1 },
    { x: 1, y: 1 },
    { x: 0, y: 2 },
    { x: 2, y: 2 },
  ]),
  new MatrixSelection([
    { x: 0, y: 0 },
    { x: 0, y: 0 },
    { x: 0, y: 1 },
    { x: 1, y: 1 },
  ]),
])('returns matrix selection length', (selection) => {
  expect(selection.length).toMatchSnapshot();
});
