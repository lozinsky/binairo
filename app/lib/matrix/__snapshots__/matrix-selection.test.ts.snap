// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`collects matrix selection from cells 1`] = `
[
  1,
  1,
  4,
  5,
  7,
  9,
]
`;

exports[`collects matrix selection from cells 2`] = `
[
  1,
  1,
  4,
  5,
]
`;

exports[`concatenates matrix selection 1`] = `
[
  {
    "x": 0,
    "y": 0,
  },
  {
    "x": 0,
    "y": 0,
  },
  {
    "x": 0,
    "y": 1,
  },
  {
    "x": 1,
    "y": 1,
  },
  {
    "x": 0,
    "y": 2,
  },
  {
    "x": 2,
    "y": 2,
  },
]
`;

exports[`concatenates matrix selection 2`] = `
[
  {
    "x": 0,
    "y": 0,
  },
  {
    "x": 0,
    "y": 0,
  },
  {
    "x": 0,
    "y": 1,
  },
  {
    "x": 1,
    "y": 1,
  },
]
`;

exports[`excludes matrix selection 1`] = `
[
  {
    "x": 0,
    "y": 0,
  },
  {
    "x": 1,
    "y": 1,
  },
]
`;

exports[`excludes matrix selection 2`] = `
[
  {
    "x": 1,
    "y": 1,
  },
]
`;

exports[`excludes matrix selection 3`] = `[]`;

exports[`returns matrix selection cells 1`] = `
[
  1,
  1,
  4,
  5,
  7,
  9,
]
`;

exports[`returns matrix selection cells 2`] = `
[
  1,
  1,
  4,
  5,
]
`;

exports[`returns matrix selection length 1`] = `6`;

exports[`returns matrix selection length 2`] = `4`;

exports[`returns matrix selection value of 1`] = `
[
  {
    "x": 0,
    "y": 0,
  },
  {
    "x": 0,
    "y": 0,
  },
  {
    "x": 0,
    "y": 1,
  },
  {
    "x": 1,
    "y": 1,
  },
  {
    "x": 0,
    "y": 2,
  },
  {
    "x": 2,
    "y": 2,
  },
]
`;

exports[`returns matrix selection value of 2`] = `
[
  {
    "x": 0,
    "y": 0,
  },
  {
    "x": 0,
    "y": 0,
  },
  {
    "x": 0,
    "y": 1,
  },
  {
    "x": 1,
    "y": 1,
  },
]
`;
